rfcs/20190815-tfdbg-v2-callbacks.md

```diff
@@ -0,0 +1,356 @@
+[`tf.function`](https://www.tensorflow.org/beta/guide/autograph) in TensorFlow 2
+("TF2" hereafter). This will enable users to "hook into" the following types of
+events in TF2 with a unified API:
+
+  - Eager execution of an operation ("op" hereafter) at runtime
+  - Creation of an symbolic op at graph-construction time, i.e., when a
+    user-defined Python functions is transformed into a graph (FuncGraph) with
+    the
+    [tf.function](https://www.tensorflow.org/beta/tutorials/eager/tf_function)
+    API.
+  - Runtime execution of FuncGraphs.
+
+## Motivation
+
+Such "hooks" will allow both observation and overriding of TF op's outgoing
+tensors, including concrete `EagerTensor`s and symbolic `Tensor`s (see details
+below.) This is a foundational part of the effort to bring [TensorFlow Debugger
+(tfdbg)](https://www.tensorflow.org/guide/debugger) up-to-date with TF2's
+execution paradigm.
+
+Currently, tfdbg is compatible with only the `tf.Session` API of TensorFlow 1.x.
+However, users of TF2 have raised questions and issues that indicate needs for
+a dedicated debugger, ones that cannot be easily met by a generic Python
+debugger such as pdb.  The most common examples of such needs involve
+finding the source of numeric instability issues like NaNs (e.g., see [GitHub
+issue](https://github.com/tensorflow/tensorflow/issues/26543) and [StackOverflow
+question](https://stackoverflow.com/questions/55823557/converting-darknet53-gives-nan-results-in-tensorflow-2-0))
+
+## Design Proposal
+
+To expand on the aims listed in the Object section, the proposed API will enable
+three key capabilities:
+
+  - **Capability A**: The ability to intercept eagerly-executing TF ops.
+    Specifically, we want a way to register one or more callbacks that are
+    invoked immediately after an `EagerTensor` has been computed through the
+    eager execution of an op. The callback should provide visibility into the
+    input and output `EagerTensor`s of the execution. If so desired, the
+    callback may _override_ the output `EagerTensor`s and thereby transparently
+    affect downstream eager execution.
+  - **Capability B**: The ability to intercept the creation of symbolic ops
+    during function-to-graph conversion in `tf.function`, including the cases
+    assisted by [AutoGraph](https://www.tensorflow.org/guide/autograph). This
+    will form the basis for simulated stepping through lines of
+    `tf.function`-decorated Python functions to assist debugging of graph
+    construction in TF2. Like in Capability A, the callbacks should be able to
+    override the output symbolic tensors of the op in order to affect all
+    downstream graph ops.
+  - **Capability C**: Similar to Capability A above, we want the ability
+    to intercept the runtime execution of FuncGraphs. Note that this
+    requirement could be folded into Capability A, if FuncGraphs are
+    regarded as a special type of op.
```

Comment:
To clarify, this means that for a FuncGraph generated by a tf.function, for example, you could inspect as if it were a single op, eg override the output tensors? And this would happen for the FuncGraph itself, rather than the call sites independently? Or am I misunderstanding the mode of operation here?